integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/massindexing/IndexWorkExecutorIT.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.integrationtest.backend.tck.massindexing;
+
+import static org.hibernate.search.util.impl.integrationtest.common.assertion.SearchResultAssert.assertThat;
+import static org.hibernate.search.util.impl.integrationtest.common.stub.mapper.StubMapperUtils.referenceProvider;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.hibernate.search.engine.backend.document.DocumentElement;
+import org.hibernate.search.engine.backend.document.IndexFieldAccessor;
+import org.hibernate.search.engine.backend.document.model.dsl.IndexSchemaElement;
+import org.hibernate.search.engine.backend.index.spi.IndexDocumentWorkExecutor;
+import org.hibernate.search.engine.backend.index.spi.IndexWorkExecutor;
+import org.hibernate.search.engine.search.DocumentReference;
+import org.hibernate.search.engine.search.SearchQuery;
+import org.hibernate.search.integrationtest.backend.tck.util.rule.SearchSetupHelper;
+import org.hibernate.search.util.impl.integrationtest.common.stub.mapper.StubMappingIndexManager;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/**
+ * Verify that the work executor operations:
+ * {@link IndexWorkExecutor#optimize()}, {@link IndexWorkExecutor#purge(String)}, {@link IndexWorkExecutor#flush()}
+ * work properly, in every backends.
+ */
+public class IndexWorkExecutorIT {
+
+	private static final String INDEX_NAME = "lordOfTheRingsChapters";
+	public static final int NUMBER_OF_BOOKS = 200;
+
+	@Rule
+	public SearchSetupHelper setupHelper = new SearchSetupHelper();
+
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
+
+	private IndexAccessors indexAccessors;
+	private StubMappingIndexManager indexManager;
+
+	@Before
+	public void before() {
+		setupHelper.withDefaultConfiguration()
+				.withIndex(
+						"MappedType", INDEX_NAME,
+						ctx -> this.indexAccessors = new IndexAccessors( ctx.getSchemaElement() ),
+						indexManager -> this.indexManager = indexManager
+				)
+				.setup();
+
+		createBookIndexes();
+		assertBookNumberIsEqualsTo( NUMBER_OF_BOOKS );
+	}
+
+	@Test
+	public void runOptimizePurgeAndFlushInSequence() {
+		IndexWorkExecutor workExecutor = indexManager.createWorkExecutor();
+
+		workExecutor.optimize().join();
+		assertBookNumberIsEqualsTo( NUMBER_OF_BOOKS );
+
+		workExecutor.purge( null ).join();
```

Comment:
You should probably also test this operation with a tenant ID, see if it properly only deletes documents for that tenant ID, and leave the other documents untouched.

You can add that test in `org.hibernate.search.integrationtest.backend.tck.MultiTenancyIT`. If you think it makes more sense to test it here, in a separate method with a separate setup, let's talk about it first.

Reply:
Do you mean having a test with 2 different tenants and verify that the purge will affect only one of them?

Reply:
Yes.